StylePropertyMap.set() should wrap value in a calc() if outside the allowed range

cdumez opened the pull request:
#### 2a77eefbdefdfe3b450259a8b108f6a79f457d7a
<pre>
StylePropertyMap.set() should wrap value in a calc() if outside the allowed range
<a href="https://bugs.webkit.org/show_bug.cgi?id=248547">https://bugs.webkit.org/show_bug.cgi?id=248547</a>

Reviewed by Antti Koivisto.

StylePropertyMap.set() should wrap value in a calc() if outside the allowed
range for the given property, instead of throwing.

This is the behavior expected by WPT tests and it matches what Blink is doing,
though I cannot find a reference to this in the specification:
- <a href="https://github.com/w3c/css-houdini-drafts/issues/1081">https://github.com/w3c/css-houdini-drafts/issues/1081</a>

This is covered by a lot of WPT tests. Sadly, those WPT tests still don&apos;t pass
because we fail a following subtest (which I&apos;ll investigate next).

For clarity, I extracted this separate check into its own test:
- fast/css/css-typed-om/style-property-map-set-negative-value.html

* LayoutTests/fast/css/css-typed-om/style-property-map-set-negative-value-expected.txt: Added.
* LayoutTests/fast/css/css-typed-om/style-property-map-set-negative-value.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/block-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/column-count-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/column-rule-width-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/flex-grow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/flex-shrink-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-size-adjust-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-stretch-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/height-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/inline-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/logical-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/orphans-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/padding-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/radius-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/shape-margin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/tab-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/widows-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/width-expected.txt:
* Source/WebCore/css/calc/CSSCalcExpressionNode.h:
(WebCore::CSSCalcExpressionNode::shouldForceEnclosingCalcInCSSText const):
(WebCore::CSSCalcExpressionNode::setShouldForceEnclosingCalcInCSSText):
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::buildCSSText):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isSimpleLengthPropertyID):
(WebCore::parseSimpleLengthValue):
(WebCore::isSimpleLengthPropertyID): Deleted.
* Source/WebCore/css/parser/CSSParserFastPaths.h:
* Source/WebCore/css/typedom/CSSStyleValue.h:
(WebCore::CSSStyleValue::toCSSValueWithProperty const):
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
(WebCore::isValueOutOfRangeForProperty):
(WebCore::CSSUnitValue::toCSSValueWithProperty const):
* Source/WebCore/css/typedom/CSSUnitValue.h:
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::cssValueFromStyleValues):

Canonical link: <a href="https://commits.webkit.org/257485@main">https://commits.webkit.org/257485@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05c67dbb7aa7616467cd9a78d34b6317ed871d8a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99051 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8254 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108446 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168695 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103013 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8799 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85615 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91563 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106400 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104775 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6663 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90243 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33678 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88485 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21588 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76533 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2150 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23104 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2048 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45514 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/7013 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42579 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2614 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3465 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on LayoutTests/fast/css/css-typed-om/style-property-map-set-negative-value.html:
```diff
@@ -0,0 +1,14 @@
+shouldBeEqualToString("target.style.width", "calc(-3.14%)");
+</script>
+</body>
+</html>
```
This should be a new WPT test.

cdumez replied:
It is already covered by the WPT tests. I am merely extracting it as a WebKit test to make the progression clear, since the WPT tests cover more than this case and now fail later on.

graouts commented on Source/WebCore/css/typedom/CSSUnitValue.cpp:
```diff
@@ -186,6 +189,59 @@ RefPtr<CSSValue> CSSUnitValue::toCSSValue() const
     return CSSPrimitiveValue::create(m_value, m_unit);
 }
 
+static bool isValueOutOfRangeForProperty(CSSPropertyID propertyID, double value, CSSUnitType unit)
```
This doesn't look exhaustive. For instance `perspective` is not in here. I guess this was informed by the set of properties used in existing WPT tests?

cdumez replied:
No, this wasn't informed by WPT tests. This matches exactly what Chrome is doing as I couldn't find a spec for this.

anttijk commented on Source/WebCore/css/typedom/CSSUnitValue.cpp:
```diff
@@ -186,6 +189,60 @@ RefPtr<CSSValue> CSSUnitValue::toCSSValue() const
+RefPtr<CSSValue> CSSUnitValue::toCSSValueWithProperty(CSSPropertyID propertyID) const
+{
+    if (isValueOutOfRangeForProperty(propertyID, m_value, m_unit)) {
+        // Wrap out of range values with a calc.
+        auto node = toCalcExpressionNode();
+        ASSERT(node);
+        // By default, cssText() will drop the enclosing `calc()` for simple expressions. However, it is important we
+        // keep it here or setting the value will fail.
+        node->setShouldForceEnclosingCalcInCSSText();
+        return CSSCalcValue::create(node.releaseNonNull(), true /* allowsNegativePercentage */);
```
This is pretty weird. It would be good to understand why we need behavior (hack?) like this. That Chrome does it is not really a sufficient reason.

cdumez committed:
StylePropertyMap.set() should wrap value in a calc() if outside the allowed range
https://bugs.webkit.org/show_bug.cgi?id=248547

Reviewed by Antti Koivisto.

StylePropertyMap.set() should wrap value in a calc() if outside the allowed
range for the given property, instead of throwing.

This is the behavior expected by WPT tests and it matches what Blink is doing,
though I cannot find a reference to this in the specification:
- https://github.com/w3c/css-houdini-drafts/issues/1081

This is covered by a lot of WPT tests. Sadly, those WPT tests still don't pass
because we fail a following subtest (which I'll investigate next).

For clarity, I extracted this separate check into its own test:
- fast/css/css-typed-om/style-property-map-set-negative-value.html

* LayoutTests/fast/css/css-typed-om/style-property-map-set-negative-value-expected.txt: Added.
* LayoutTests/fast/css/css-typed-om/style-property-map-set-negative-value.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/block-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/column-count-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/column-rule-width-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/flex-grow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/flex-shrink-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-size-adjust-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-stretch-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/height-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/inline-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/logical-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/orphans-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/padding-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/radius-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/shape-margin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/tab-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/widows-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/width-expected.txt:
* Source/WebCore/css/calc/CSSCalcExpressionNode.h:
(WebCore::CSSCalcExpressionNode::shouldForceEnclosingCalcInCSSText const):
(WebCore::CSSCalcExpressionNode::setShouldForceEnclosingCalcInCSSText):
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::buildCSSText):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isSimpleLengthPropertyID):
(WebCore::parseSimpleLengthValue):
(WebCore::isSimpleLengthPropertyID): Deleted.
* Source/WebCore/css/parser/CSSParserFastPaths.h:
* Source/WebCore/css/typedom/CSSStyleValue.h:
(WebCore::CSSStyleValue::toCSSValueWithProperty const):
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
(WebCore::isValueOutOfRangeForProperty):
(WebCore::CSSUnitValue::toCSSValueWithProperty const):
* Source/WebCore/css/typedom/CSSUnitValue.h:
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::cssValueFromStyleValues):

Canonical link: https://commits.webkit.org/257485@main
